Disallows `#![feature(no_coverage)]` on stable and beta (using standard crate-level gating)

richkadel opened the pull request:
Fixes: #84836

Removes the function-level feature gating solution originally implemented, and solves the same problem using `allow_internal_unstable`, so normal crate-level feature gating mechanism can still be used (which disallows the feature on stable and beta).

I tested this, building the compiler with and without `CFG_DISABLE_UNSTABLE_FEATURES=1`

With unstable features disabled, I get the expected result as shown here:

```shell
$ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc     src/test/run-make-fulldeps/coverage/no_cov_crate.rs
error[E0554]: `#![feature]` may not be used on the dev release channel
 --> src/test/run-make-fulldeps/coverage/no_cov_crate.rs:2:1
  |
2 | #![feature(no_coverage)]
  | ^^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to previous error

For more information about this error, try `rustc --explain E0554`.
```

r? @Mark-Simulacrum 
cc: @tmandry @wesleywiser

richkadel commented:
Actually, I'm changing the error span to point to the `feature`, not the attribute. Testing this right now and I'll upload a small change to this in a few minutes.

richkadel commented:
Fixed:

```shell
$ ./build/x86_64-unknown-linux-gnu/stage1/bin/rustc     src/test/run-make-fulldeps/coverage/no_cov_func.rs 
error[E0554]: `#[feature(no_coverage)]` may not be used on the dev release channel
 --> src/test/run-make-fulldeps/coverage/no_cov_func.rs:3:1
  |
3 | #[feature(no_coverage)]
  | ^^^^^^^^^^^^^^^^^^^^^^^

error[E0554]: `#[feature(no_coverage)]` may not be used on the dev release channel
  --> src/test/run-make-fulldeps/coverage/no_cov_func.rs:10:1
   |
10 | #[feature(no_coverage)]
   | ^^^^^^^^^^^^^^^^^^^^^^^

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0554`.
```

richkadel commented:
Oh, I also just noticed I should be using `tcx.features()` instead of `tcx.sess.features_untracked()`, so I verified this still works with that change and amended my commit.

hyd-dev commented:
This does not seem like what's described in https://github.com/rust-lang/rust/issues/84836#issuecomment-831178222 (emphasis mine)?
> It should be possible to adjust the **normal** feature gating to correctly handle macro expanded sources - that is, allow the built-in derives to generate it without it being stabilized. We do this for **the Structural trait impls** somehow, for example; possibly enough to set the span appropriately.

The macro generates `impl` for feature-gated [`StructuralEq`](https://doc.rust-lang.org/nightly/core/marker/trait.StructuralEq.html) without expanding to any `#[feature]`: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7d5fca8b78b2ce2dcc18c750d60af69a (click "TOOLS" -> "Expand macros")

hyd-dev commented:
This does not seem like what's described in https://github.com/rust-lang/rust/issues/84836#issuecomment-831178222 (emphasis mine)?
> It should be possible to adjust the **normal** feature gating to correctly handle macro expanded sources - that is, allow the built-in derives to generate it without it being stabilized. We do this for **the Structural trait impls** somehow, for example; possibly enough to set the span appropriately.

The macro generates `impl` for feature-gated [`StructuralEq`](https://doc.rust-lang.org/nightly/core/marker/trait.StructuralEq.html) without expanding to any `#[feature]`: https://play.rust-lang.org/?version=nightly&mode=debug&edition=2018&gist=7d5fca8b78b2ce2dcc18c750d60af69a (click "TOOLS" -> "Expand macros")

richkadel commented:
> This does not seem like what's described in [#84836 (comment)](https://github.com/rust-lang/rust/issues/84836#issuecomment-831178222) (emphasis mine)?
> 
> >

No, probably not.  I don't fully understand the option Mark described, and I'm not sure if it's a general solution, or one that will only work for the `core` library.

But this solution has the desired effect.

tmandry commented:
This will work for now but given https://github.com/rust-lang/rust/issues/84836#issuecomment-832303085 I don't think we should continue using this mechanism without an MCP.

This wouldn't work anyway if we decided to stabilize `-Zinstrument-coverage` without the `no_coverage` feature.

tmandry commented on compiler/rustc_typeck/src/collect.rs:
```diff
@@ -2730,7 +2730,7 @@ fn codegen_fn_attrs(tcx: TyCtxt<'_>, id: DefId) -> CodegenFnAttrs {
                 if list.iter().any(|nested_meta_item| nested_meta_item.has_name(sym::no_coverage)) {
                     tcx.sess.mark_attr_used(attr);
-                    no_coverage_feature_enabled = true;
+                    no_coverage_feature_attr = Some(attr);
```
The nightly check would need to happen here, btw. Otherwise we'd allow `#[feature(no_coverage)]` without `#[no_coverage]`.

richkadel replied:
You're right. Thanks. 

Nevertheless, this is now outdated.

richkadel commented:
I just uploaded a revised implementation that I believe meets everyone's expectations. Per @nagisa 's hints, I figured out how to apply the feature gating at the crate level, and used `allow_internal_unstable` when injecting the derived method from `Eq`.

I updated the `ui` test to include a struct that derives `Eq` to make sure it still fails on `#[no_coverage]` applied elsewhere.

Tests with `#[no_coverage]` that also enable the feature still work on nightly, and they still fail on beta/stable (validated using the same method described in this PR description above).

nagisa commented on compiler/rustc_builtin_macros/src/deriving/cmp/eq.rs:
```diff
@@ -15,20 +15,15 @@ pub fn expand_deriving_eq(
-        cx.attribute(doc),
-    ];
+    let mut attrs = vec![cx.attribute(inline), cx.attribute(doc)];
+    if cx.parse_sess().unstable_features.is_nightly_build() {
```
This doesn't need to be conditional on `is_nightly_build`, does it?

richkadel replied:
I'll try to change it and verify it still builds `core` and still disables use on beta/stable.

richkadel replied:
Done

nagisa commented:
Can you adjust the description and the title of the PR please?

richkadel commented:
> Can you adjust the description and the title of the PR please?

I added an exclamation point to the attribute in the title.

The description looks fine to me. (I had already removed the function-level example.)

Let me know if I'm missing something.

richkadel commented:
@nagisa - I think this is ready:

* I addressed your comment in eq.rs and removed the conditional (verified it still blocks on beta/stable, and it still works otherwise)
* I updated the PR title, and the commit message. I had already updated the PR commit message last night.
* I squashed the commits

Thanks

nagisa commented:
> Let me know if I'm missing something.

Ah, I meant that its no longer precisely about disallowing the attribute on stable/beta (in particular no code specifically addresses this particular point) but more like about making `no_coverage` use the usual feature gating infrastructure. But I don't care too strongly either way.

@bors r+

bors commented:
:pushpin: Commit 3584c1dd0cb501916c8bc6fd452864b261068beb has been approved by `nagisa`

<!-- @bors r=nagisa 3584c1dd0cb501916c8bc6fd452864b261068beb -->
<!-- homu: {"type":"Approved","sha":"3584c1dd0cb501916c8bc6fd452864b261068beb","approver":"nagisa"} -->

richkadel commented:
Oh, I gotcha. Done.

pietroalbini commented:
My understanding is that this should be backported to 1.53 beta, correct?

wesleywiser commented:
Yes, nominating for beta-backport.

pnkfelix commented:
discussed in T-compiler meeting, beta backport approved.
